tests/robustness/model/model.go

```diff
@@ -33,469 +30,92 @@ const (
-		return "nil"
-	}
-	return fmt.Sprintf("%q", value.Value)
-}
-
-func step(states PossibleStates, request EtcdRequest, response EtcdResponse) (bool, PossibleStates) {
+func (states nonDeterministicState) Step(request EtcdRequest, response EtcdNonDeterministicResponse) (bool, nonDeterministicState) {
 	if len(states) == 0 {
 		// states were not initialized
 		if response.Err != nil || response.ResultUnknown || response.Revision == 0 {
 			return true, nil
 		}
-		return true, PossibleStates{initState(request, response)}
+		return true, initNonDeterministicState(request, response)
 	}
+	var newStates nonDeterministicState
 	if response.Err != nil {
-		states = applyFailedRequest(states, request)
+		newStates = states.stepFailedRequest(request)
 	} else {
-		states = applyRequest(states, request, response)
+		newStates = states.stepSuccessfulRequest(request, response)
 	}
-	return len(states) > 0, states
+	return len(newStates) > 0, newStates
 }
 
-// initState tries to create etcd state based on the first request.
-func initState(request EtcdRequest, response EtcdResponse) EtcdState {
-	state := EtcdState{
-		Revision:  response.Revision,
-		KeyValues: map[string]ValueRevision{},
-		KeyLeases: map[string]int64{},
-		Leases:    map[int64]EtcdLease{},
-	}
-	switch request.Type {
-	case Txn:
-		if response.Txn.TxnResult {
-			return state
-		}
-		if len(request.Txn.Ops) != len(response.Txn.OpsResult) {
-			panic(fmt.Sprintf("Incorrect request %s, response %+v", describeEtcdRequest(request), describeEtcdResponse(request, response)))
-		}
-		for i, op := range request.Txn.Ops {
-			opResp := response.Txn.OpsResult[i]
-			switch op.Type {
-			case Range:
-				for _, kv := range opResp.KVs {
-					state.KeyValues[kv.Key] = ValueRevision{
-						Value:       kv.Value,
-						ModRevision: kv.ModRevision,
-					}
-				}
-			case Put:
-				state.KeyValues[op.Key] = ValueRevision{
-					Value:       op.Value,
-					ModRevision: response.Revision,
-				}
-			case Delete:
-			default:
-				panic("Unknown operation")
-			}
-		}
-	case LeaseGrant:
-		lease := EtcdLease{
-			LeaseID: request.LeaseGrant.LeaseID,
-			Keys:    map[string]struct{}{},
-		}
-		state.Leases[request.LeaseGrant.LeaseID] = lease
-	case LeaseRevoke:
-	case Defragment:
-	default:
-		panic(fmt.Sprintf("Unknown request type: %v", request.Type))
-	}
-	return state
+func initNonDeterministicState(request EtcdRequest, response EtcdNonDeterministicResponse) nonDeterministicState {
+	return nonDeterministicState{initState(request, response.EtcdResponse)}
 }
 
-// applyFailedRequest handles a failed requests, one that it's not known if it was persisted or not.
-func applyFailedRequest(states PossibleStates, request EtcdRequest) PossibleStates {
+// stepFailedRequest duplicates number of states by considering request persisted and lost.
+func (states nonDeterministicState) stepFailedRequest(request EtcdRequest) nonDeterministicState {
+	newStates := make(nonDeterministicState, 0, len(states)*2)
```

Comment:
you are doubling number of states, is there a scenario that the number will grow very quickly and checker won't be able to handle it?

Reply:
I guess `stepSuccessfulRequest` would remove possible states. One successful response should clean up all previous possible states generated back to back.

Reply:
You are both right. NonDeterministic model can grow exponentially with number of operations. This however mostly happens on failed/error (wanna guess when it can happen on request success?). As @chaochn47 mentioned successful requests will cleanup previous possible states. 

So the problem is that we should avoid long sequences of failed requests as they cause complexity to explode. I have mitigated the problem by introducing history patching https://github.com/etcd-io/etcd/pull/15078 https://github.com/etcd-io/etcd/issues/15598